lib/utils/remote-build.ts

```diff
@@ -0,0 +1,213 @@
+				input: process.stdin,
+				output: process.stdout,
+			});
+
+			readline.on('SIGINT', () => process.emit('SIGINT'));
+		}
+		//
+		// Setup interrupt handlers so we can cancel the build if the user presses
+		// ctrl+c
+
+		// This is necessary because the `exit-hook` module is used by several
+		// dependencies, and will exit without calling the following handler.
+		// Once https://github.com/resin-io/resin-cli/issues/867 has been solved,
+		// we are free to (and definitely should) remove the below line
+		process.removeAllListeners('SIGINT');
+		process.on('SIGINT', () => {
+			process.stderr.write('Received SIGINT, cleaning up. Please wait.\n');
+			cancelBuildIfNecessary(build).then(() => {
+				stream.end();
+				process.exit(0);
+			});
+		});
+
+		stream.on('data', getBuilderMessageHandler(build));
+		stream.on('end', resolve);
+		stream.on('error', reject);
+	}).then(() => {
+		build.running = false;
+	});
+}
+
+async function handleBuilderMetadata(obj: BuilderMessage, build: RemoteBuild) {
+	switch (obj.resource) {
+		case 'cursor':
+			const readline = await import('readline');
+
+			if (obj.value == null) {
+				return;
+			}
+
+			const match = obj.value.match(CURSOR_METADATA_REGEX);
+
+			if (!match) {
+				// FIXME: Make this error nicer.
+				console.log(
+					'Warning: ignoring unknown builder output. You may experience odd build output. ' +
+						'Maybe you need to update resin-cli?',
+				);
+				return;
+			}
+
+			const value = match[1];
+			const amount = match[2] || 1;
+
+			switch (value) {
+				case 'erase':
+					readline.clearLine(process.stdout, 0);
+					process.stdout.write('\r');
+					break;
+				case 'up':
+					readline.moveCursor(process.stdout, 0, -amount);
+					break;
+				case 'down':
+					readline.moveCursor(process.stdout, 0, amount);
+					break;
+			}
+
+			break;
+		case 'buildLogId':
+			// The name of this resource is slightly dated, but this is the release
+			// id from the API. We need to save this so that if the user ctrl+c's the
+			// build we can cancel it on the API.
+			build.releaseId = parseInt(obj.value!, 10);
+			break;
+	}
+}
+
+function getBuilderMessageHandler(
+	build: RemoteBuild,
+): (obj: BuilderMessage) => Promise<void> {
+	return async (obj: BuilderMessage) => {
+		if (DEBUG_MODE) {
+			console.log(`[debug] handling message: ${JSON.stringify(obj)}`);
+		}
+		if (obj.type != null && obj.type === 'metadata') {
+			return handleBuilderMetadata(obj, build);
+		}
+		if (obj.message) {
+			const readline = await import('readline');
+			readline.clearLine(process.stdout, 0);
+
+			if (!SUPPRESS_OUTPUT) {
+				const message = obj.message.replace(TRIM_REGEX, '');
+				if (obj.replace) {
+					process.stdout.write(`\r${message}`);
+				} else {
+					console.log(message);
+				}
+			}
+		}
+	};
+}
+
+async function cancelBuildIfNecessary(build: RemoteBuild): Promise<void> {
+	if (build.running && build.releaseId != null) {
```

Comment:
Why do we need the separate `running` flag? Can't we just check if we've got a release id?